error.go

```diff
@@ -475,3 +493,114 @@ func AppendInto(into *error, err error) (errored bool) {
 	*into = Append(*into, err)
 	return true
 }
+
+// Invoker is an operation that may fail with an error. Use it with
+// AppendInvoke to append the result of calling the function into a function.
+// This allows you to conveniently defer capture of failing operations.
+//
+// See also, Close and Invoke.
+type Invoker interface {
+	Invoke() error
+}
+
+// Invoke wraps a function which may fail with an error to match the Invoker
+// interface. Use it to supply functions matching this signature to
+// AppendInvoke.
+//
+// For example,
+//
+// 	func processReader(r io.Reader) (err error) {
+// 		scanner := bufio.NewScanner(r)
+// 		defer multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))
+// 		for scanner.Scan() {
+// 			// ...
+// 		}
+// 		// ...
+// 	}
+//
+// In this example, the following line will construct the Invoker right away,
+// but defer the invocation of scanner.Err() until the function returns.
+//
+// 	defer multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))
+type Invoke func() error
```

Comment:
I like the function --> interface wrapper, but it does lead to a little bit of stutter (`muiltierr.AppendInvoke` and  `multierr.Invoke`), so just want to compare with adding an additional top-level method instead like `AppendInvokeFn`,
```
# current
multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))

# a little less stutter, but additional API
multierr.AppendInvokeFn(&err, scanner.Err)
```

Generally it's better to only have one way of doing something, this may be a common enough operation to add both for. This is also similar to how the http package supports both `HandlerFunc` type but also has `HandleFunc` methods at the package/servemux.

Reply:
I'm open for both as I originally proposed something like the `AppenInvokeFn` you're suggesting now.

I do not really feel like it would 'stutter' thinking of always creating an `Invoke` that is going to be appended therefore it is somewhat consequent but on the other hand I think you have a point that having something like `func() error` is a very common use case.

Maybe @abhinav has a 'strong opinion' about this?
The `multierr.Close` wrapper was my personal compromise between the interface approach and being able to use this in an convenient way but I would also be fine having something like `AppendInvokeErrFunc(err *error, func() error)` and remove the `multierr.Close` wrapper at all.

But I'd like to mention that we are going around in circles here considering the discussion in the original issue.

Reply:
I don't have a strong opinion on AppendInvokeFn but I would lean towards not including AppendInvokeFn initially because I'm not a fan of the name. We can add something like that in the future if enough need arises.

I agree that there's some repetition in,

    multierr.AppendInvoke(&err, multierr.Invoke(scanner.Err))

But missing a more obvious, shorter name, this seems like an acceptable cost.